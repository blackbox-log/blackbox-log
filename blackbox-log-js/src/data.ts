import { Memoize as memoize } from 'typescript-memoize';

import { WasmPointer } from './wasm';

import type { Headers } from './headers';
import type { WasmExports, WasmObject } from './wasm';

export class DataParser implements WasmObject {
	readonly #wasm: WasmExports;
	readonly #ptr: WasmPointer;
	readonly #headers: Headers;

	constructor(wasm: WasmExports, headers: Headers, ptr: number) {
		this.#wasm = wasm;
		this.#headers = headers;
		this.#ptr = new WasmPointer(ptr, wasm.data_free);
	}

	free() {
		this.#ptr.free();
	}

	get isAlive(): boolean {
		return this.#ptr.isAlive;
	}

	get headers(): Headers {
		return this.#headers;
	}

	parseAll() {
		this.#wasm.data_parseAll(this.#ptr.ptr);
	}

	get mainFrameCount(): number {
		return this.#wasm.data_mainFrameCount(this.#ptr.ptr);
	}

	get gpsFrameCount(): number {
		return this.#wasm.data_gpsFrameCount(this.#ptr.ptr);
	}
}
