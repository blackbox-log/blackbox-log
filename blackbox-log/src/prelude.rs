//! Minimal set of imports for convenience.
//!
//! [`File`](`crate::File`) is not included due to avoid collisions with other
//! types.

pub use crate::frame::{Frame as _, FrameDef as _};
pub use crate::{DataParser, FieldFilter, Headers, ParserEvent};
