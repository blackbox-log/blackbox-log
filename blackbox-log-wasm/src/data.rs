use blackbox_log::prelude::*;
use blackbox_log::Reader;

use crate::{OwnedSlice, Shared, WasmFfi};

// SAFETY: field order *must* be `parser` first, then `headers`, then `data` to
// ensure correct drop order
pub struct WasmDataParser {
    parser: DataParser<'static, 'static>,
    _headers: Shared<Headers<'static>>,
    _data: Shared<OwnedSlice>,
}

impl WasmDataParser {
    pub(crate) fn new(
        headers: Shared<Headers<'static>>,
        reader: Reader<'static>,
        data: Shared<OwnedSlice>,
    ) -> Self {
        // SAFETY: this is only used to create the `DataParser`, which is guaranteed to
        // be dropped before `headers` by the declaration order in the struct
        let headers_ref = unsafe { headers.deref_static() };

        Self {
            parser: DataParser::new(reader, headers_ref),
            _headers: headers,
            _data: data,
        }
    }

    fn main_frame_count(&self) -> usize {
        self.parser.stats().counts.main
    }

    fn gps_frame_count(&self) -> usize {
        self.parser.stats().counts.gps
    }
}

impl WasmFfi for WasmDataParser {}

#[no_mangle]
unsafe extern "wasm" fn data_free(ptr: *mut WasmDataParser) {
    let parser = WasmDataParser::from_wasm(ptr);
    drop(parser);
}

#[no_mangle]
#[allow(non_snake_case)]
unsafe extern "wasm" fn data_mainFrameCount(ptr: *mut WasmDataParser) -> usize {
    let parser = WasmDataParser::from_wasm(ptr);
    let count = parser.main_frame_count();
    parser.into_wasm();
    count
}

#[no_mangle]
#[allow(non_snake_case)]
unsafe extern "wasm" fn data_gpsFrameCount(ptr: *mut WasmDataParser) -> usize {
    let parser = WasmDataParser::from_wasm(ptr);
    let count = parser.gps_frame_count();
    parser.into_wasm();
    count
}
