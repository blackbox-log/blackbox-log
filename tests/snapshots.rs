use std::collections::BTreeMap;
use std::fs;
use std::io::Read;

use blackbox_log::data::Stats;
use blackbox_log::frame::{self, FieldDef, GpsFrameDef};
use blackbox_log::prelude::*;
use blackbox_log::units::si;
use blackbox_log::{headers, Event, Unit, Value};
use hashbrown::HashMap;
use serde::ser::{SerializeStruct, Serializer};
use serde::Serialize;

const MAX_EVENTS: usize = 200_000;
const MAX_LOGS: usize = 2;

macro_rules! run {
    () => {
        |path| {
            let mut file = fs::File::open(path).unwrap();
            let mut data = Vec::new();
            file.read_to_end(&mut data).unwrap();

            let file = blackbox_log::File::new(&data);
            let snapshot = FileSnapshot::from(file);

            insta::assert_ron_snapshot!(snapshot);
        }
    };
}

#[test]
fn own() {
    insta::glob!("logs/*.bbl", run!());
}

#[test]
#[ignore]
fn fc_blackbox() {
    insta::glob!("logs/fc-blackbox/*", run!());
}

#[test]
#[ignore]
fn gimbal_ghost() {
    insta::glob!("logs/gimbal-ghost/*", run!());
}

#[derive(Debug, Serialize)]
#[serde(rename = "File")]
struct FileSnapshot<'data> {
    count: usize,
    logs: Vec<headers::ParseResult<LogSnapshot<'data>>>,
}

impl<'data> From<blackbox_log::File<'data>> for FileSnapshot<'data> {
    fn from(file: blackbox_log::File<'data>) -> Self {
        let logs = file
            .iter()
            .map(|mut reader| {
                Headers::parse(&mut reader).map(|headers| {
                    let data = DataParser::new(reader, &headers);
                    LogSnapshot::new(&headers, data)
                })
            })
            .take(MAX_LOGS)
            .collect::<Vec<_>>();

        Self {
            count: file.log_count(),
            logs,
        }
    }
}

#[derive(Debug, Serialize)]
#[serde(rename = "Log")]
struct LogSnapshot<'data> {
    headers: HeadersSnapshot<'data>,
    stats: Stats,
    capped: bool,
    #[serde(skip_serializing_if = "Vec::is_empty")]
    events: Vec<Event>,
    #[serde(skip_serializing_if = "MainSnapshot::is_empty")]
    main: MainSnapshot,
    #[serde(skip_serializing_if = "SlowSnapshot::is_empty")]
    slow: SlowSnapshot,
    #[serde(skip_serializing_if = "GpsSnapshot::is_empty")]
    gps: GpsSnapshot,
}

impl<'data> LogSnapshot<'data> {
    fn new(headers: &Headers<'data>, mut data: DataParser<'data, '_>) -> Self {
        let mut events = Vec::new();

        let main = headers.main_frame_def.iter().collect::<Fields>();
        let mut main = MainSnapshot::new(main);

        let slow = headers.slow_frame_def.iter().collect::<Fields>();
        let mut slow = SlowSnapshot::new(slow);

        let gps = headers
            .gps_frame_def
            .iter()
            .flat_map(GpsFrameDef::iter)
            .collect::<Fields>();
        let mut gps = GpsSnapshot::new(gps);

        let mut capped = true;
        for _ in 0..MAX_EVENTS {
            let Some(frame) = data.next() else {
                capped = false;
                break;
            };

            match frame {
                ParserEvent::Event(event) => events.push(event),
                ParserEvent::Main(frame) => main.update(frame),
                ParserEvent::Slow(frame) => slow.update(frame),
                ParserEvent::Gps(frame) => gps.update(frame),
            }
        }

        Self {
            headers: headers.into(),
            stats: data.stats().clone(),
            capped,
            events,
            main,
            slow,
            gps,
        }
    }
}

#[derive(Debug, Serialize)]
#[serde(rename = "Headers")]
struct HeadersSnapshot<'data> {
    firmware_revision: &'data str,
    firmware: headers::Firmware,
    firmware_date: Option<Result<String, &'data str>>,
    board_info: Option<&'data str>,
    craft_name: Option<&'data str>,
    debug_mode: headers::DebugMode,
    disabled_fields: headers::DisabledFields,
    features: headers::FeatureSet,
    unknown: HashMap<&'data str, &'data str>,
}

impl<'data, 'a> From<&'a Headers<'data>> for HeadersSnapshot<'data> {
    fn from(headers: &'a Headers<'data>) -> Self {
        HeadersSnapshot {
            firmware_revision: headers.firmware_revision(),
            firmware: headers.firmware(),
            firmware_date: headers
                .firmware_date()
                .map(|res| res.map(|date| date.to_string())),
            board_info: headers.board_info(),
            craft_name: headers.craft_name(),
            debug_mode: headers.debug_mode(),
            disabled_fields: headers.disabled_fields(),
            features: headers.features(),
            unknown: headers.unknown().clone(),
        }
    }
}

#[derive(Debug, Serialize)]
#[serde(rename = "MainFrame")]
struct MainSnapshot {
    count: u128,
    last_iteration: u32,
    time: NumberHistory<16, i128>,
    fields: Fields,
}

impl MainSnapshot {
    fn new(fields: Fields) -> Self {
        Self {
            count: 0,
            last_iteration: 0,
            time: NumberHistory::new(),
            fields,
        }
    }

    fn update(&mut self, frame: frame::MainFrame) {
        self.count += 1;
        self.last_iteration = frame.iteration();
        self.time.update(frame.time_raw().into());
        self.fields.update(frame);
    }

    fn is_empty(&self) -> bool {
        self.count == 0
    }
}

#[derive(Debug, Serialize)]
#[serde(rename = "SlowFrame")]
struct SlowSnapshot {
    count: u32,
    fields: Fields,
}

impl SlowSnapshot {
    fn new(fields: Fields) -> Self {
        Self { count: 0, fields }
    }

    fn update(&mut self, frame: frame::SlowFrame) {
        self.count += 1;
        self.fields.update(frame);
    }

    fn is_empty(&self) -> bool {
        self.count == 0
    }
}

#[derive(Debug, Serialize)]
#[serde(rename = "GpsFrame")]
struct GpsSnapshot {
    count: u32,
    time: (),
    fields: Fields,
}

impl GpsSnapshot {
    fn new(fields: Fields) -> Self {
        Self {
            count: 0,
            time: (),
            fields,
        }
    }

    fn update(&mut self, frame: frame::GpsFrame) {
        self.count += 1;
        self.fields.update(frame);
    }

    fn is_empty(&self) -> bool {
        self.count == 0
    }
}

#[derive(Debug, Serialize)]
struct Fields(Vec<FieldSnapshot>);

impl Fields {
    fn update<F: blackbox_log::frame::Frame>(&mut self, frame: F) {
        for (field, value) in self.0.iter_mut().zip(frame.iter()) {
            field.update(value.into());
        }
    }
}

impl<'data, U: Into<Unit>> FromIterator<FieldDef<'data, U>> for Fields {
    fn from_iter<I: IntoIterator<Item = FieldDef<'data, U>>>(iter: I) -> Self {
        Self(
            iter.into_iter()
                .map(|field| FieldSnapshot::new(field.name.to_owned(), field.unit.into()))
                .collect(),
        )
    }
}

#[derive(Debug, Serialize)]
#[serde(rename = "Field")]
struct FieldSnapshot {
    name: String,
    unit: Unit,
    history: History,
}

#[derive(Debug, Serialize)]
#[serde(untagged)]
enum History {
    Int(NumberHistory<16, i128>),
    Bool { yes: usize, no: usize },
    Flags(BTreeMap<&'static str, usize>),
}

#[derive(Debug)]
struct NumberHistory<const N: usize, T> {
    min: T,
    max: T,
    seen: Vec<T>,
    histogram: [usize; N],
}

impl FieldSnapshot {
    fn new(name: String, unit: Unit) -> Self {
        Self {
            name,
            unit,
            history: match unit {
                Unit::Amperage
                | Unit::Voltage
                | Unit::Acceleration
                | Unit::Rotation
                | Unit::GpsCoordinate
                | Unit::Altitude
                | Unit::Velocity
                | Unit::GpsHeading
                | Unit::Unitless => History::Int(NumberHistory::new()),
                Unit::Boolean => History::Bool { yes: 0, no: 0 },
                Unit::FlightMode | Unit::State | Unit::FailsafePhase => {
                    History::Flags(BTreeMap::new())
                }
            },
        }
    }

    #[allow(clippy::wildcard_enum_match_arm, clippy::cast_possible_truncation)]
    fn update(&mut self, value: Value) {
        match &mut self.history {
            History::Int(history) => history.update(match value {
                Value::Amperage(a) => a.get::<si::electric_current::milliampere>().round() as i128,
                Value::Voltage(v) => v.get::<si::electric_potential::millivolt>().round() as i128,
                Value::Acceleration(a) => a
                    .get::<si::acceleration::centimeter_per_second_squared>()
                    .round() as i128,
                Value::Rotation(r) => {
                    r.get::<si::angular_velocity::degree_per_second>().round() as i128
                }
                Value::GpsCoordinate(c) => (c * 10000000.).round() as i128,
                Value::Altitude(a) => a.get::<si::length::meter>().round() as i128,
                Value::Velocity(v) => {
                    v.get::<si::velocity::centimeter_per_second>().round() as i128
                }
                Value::GpsHeading(h) => (h * 10.).round() as i128,
                Value::Unsigned(u) => u.into(),
                Value::Signed(s) => s.into(),
                _ => unreachable!(),
            }),
            History::Bool { yes, no } => match value {
                Value::Boolean(true) => *yes += 1,
                Value::Boolean(false) => *no += 1,
                _ => unreachable!(),
            },
            History::Flags(history) => {
                if let Value::FailsafePhase(phase) = value {
                    *history.entry(phase.as_name()).or_insert(0) += 1;
                } else {
                    let flags = match value {
                        Value::FlightMode(m) => m.as_names(),
                        Value::State(s) => s.as_names(),
                        _ => unreachable!(),
                    };

                    for flag in flags {
                        *history.entry(flag).or_insert(0) += 1;
                    }
                }
            }
        }
    }
}

impl<const N: usize, T> NumberHistory<N, T> {
    fn new() -> Self
    where
        T: Default,
    {
        Self {
            min: T::default(),
            max: T::default(),
            seen: Vec::new(),
            histogram: [0; N],
        }
    }

    fn update_range(&mut self, value: T)
    where
        T: PartialOrd,
    {
        if value < self.min {
            self.min = value;
        } else if value > self.max {
            self.max = value;
        }
    }

    /// Store new value sorted in `seen` if not already present
    fn update_seen(&mut self, value: T)
    where
        T: PartialOrd,
    {
        let index = self.seen.partition_point(|x| *x < value);
        if self.seen.get(index) != Some(&value) {
            self.seen.insert(index, value);
        }
    }

    /// Bucket so that sequential values go in sequential buckets, since data is
    /// usually clustered
    #[inline(always)]
    fn update_histogram(&mut self, value: u128) {
        self.histogram[(value % N as u128) as usize] += 1;
    }
}

impl<const N: usize> NumberHistory<N, i128> {
    fn update(&mut self, value: i128) {
        self.update_range(value);
        self.update_seen(value);
        self.update_histogram(value.unsigned_abs());
    }
}

impl<const N: usize> Serialize for NumberHistory<N, i128>
where
    [usize; N]: Serialize,
{
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        serialize(self, self.min, self.max, serializer)
    }
}

impl<const N: usize> Serialize for NumberHistory<N, f64>
where
    [usize; N]: Serialize,
{
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        serialize(
            self,
            format!("{:.2}", self.min),
            format!("{:.2}", self.max),
            serializer,
        )
    }
}

#[inline(always)]
fn serialize<const N: usize, T, S, U>(
    snapshot: &NumberHistory<N, T>,
    min: U,
    max: U,
    serializer: S,
) -> Result<S::Ok, S::Error>
where
    S: Serializer,
    U: Serialize,
    [usize; N]: Serialize,
{
    let unique = snapshot.seen.len();

    let mut state = serializer.serialize_struct("FieldSnapshot", 4)?;

    state.serialize_field("min", &min)?;
    state.serialize_field("max", &max)?;
    state.serialize_field("unique", &unique)?;

    let histogram = "histogram";
    if unique > 1 {
        state.serialize_field(histogram, &snapshot.histogram)?;
    } else {
        state.skip_field(histogram)?;
    }

    state.end()
}
