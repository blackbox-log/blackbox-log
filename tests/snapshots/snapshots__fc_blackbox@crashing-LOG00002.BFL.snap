---
source: tests/snapshots.rs
expression: snapshot
input_file: ".modules/fc-blackbox/src/test-data/crashing-LOG00002.BFL"
---
FileSnapshot(
  count: 1,
  logs: [
    Ok(LogSnapshot(
      headers: Headers(
        version: V2,
        firmware_revision: "EmuFlight 3.7.0 (9366544) OMNIBUSF4SD",
        firmware_kind: EmuFlight,
        board_info: None,
        craft_name: Some("25x25_blaxkbox_s"),
        vbat_reference: None,
        acceleration_1g: Some(2048),
        gyro_scale: Some(0.017453292),
        min_throttle: Some(1070),
        motor_output_range: Some(MotorOutputRange(
          min: 1070,
          max: 2000,
        )),
        unknown: {
          "acc_hardware": "0",
          "emu_boost_yaw": "0",
          "deadband": "0",
          "gyro_lowpass2_type": "0",
          "gyro_lowpass2_hz_yaw": "0",
          "pid_process_denom": "1",
          "gyro_lowpass2_hz_pitch": "0",
          "gyro_cal_on_first_arm": "0",
          "acc_limit_yaw": "0",
          "mag_hardware": "1",
          "spa_pitch_i": "85",
          "yaw_deadband": "0",
          "maxthrottle": "2000",
          "witchcraft_roll": "2",
          "I interval": "256",
          "dterm_lowpass_hz_roll": "90",
          "smart_smoothing_roll": "0",
          "rate_correction": "10,10",
          "gyro_sync_denom": "1",
          "rate_weight": "0,0",
          "gyro_32khz_hardware_lpf": "0",
          "gyro_lowpass2_hz_roll": "0",
          "serialrx_provider": "9",
          "levelPIDLOW": "100,10,40",
          "gyro_lowpass_hz_yaw": "0",
          "spa_pitch_d": "110",
          "spa_roll_i": "85",
          "dterm_lowpass2_hz_yaw": "200",
          "emu_boost": "0",
          "gyro_lowpass_type": "0",
          "witchcraft_yaw": "0",
          "gyro_lowpass_hz_roll": "0",
          "Log start datetime": "0000-01-01T00:00:00.000+00:00",
          "smart_smoothing_pitch": "0",
          "dterm_filter_type": "1",
          "smart_smoothing_yaw": "0",
          "gyro_lowpass_hz_pitch": "0",
          "rc_interpolation_interval": "19",
          "dterm_lowpass2_hz_roll": "200",
          "rc_interpolation": "2",
          "tpa_breakpoint": "1600",
          "dterm_lowpass_hz_pitch": "90",
          "rates": "70,70,70",
          "acc_limit": "0",
          "gyro_notch_hz": "260,0",
          "baro_hardware": "1",
          "boost_limit_yaw": "0",
          "rate_sensitivity": "100,100",
          "acc_lpf_hz": "4000",
          "nfe_racermode": "0",
          "rc_rates": "100,100,100",
          "pidAtMinThrottle": "1",
          "pitchPID": "58,70,30",
          "spa_yaw_p": "130",
          "magPID": "40",
          "P interval": "4",
          "pidsum_limit_yaw": "400",
          "spa_yaw_i": "70",
          "spa_yaw_d": "130",
          "tpa_rate_i": "125",
          "boost_limit": "0",
          "Firmware date": "May 19 2020 04:18:37",
          "pidsum_limit": "1000",
          "i_decay": "1",
          "levelPIDHIGH": "35,1",
          "iterm_rotation": "0",
          "thr_mid": "50",
          "witchcraft_pitch": "2",
          "spa_roll_p": "110",
          "throttle_boost": "0",
          "rollPID": "50,70,28",
          "thr_expo": "0",
          "feathered_pids": "100",
          "looptime": "125",
          "yawPID": "60,70,5",
          "spa_pitch_p": "110",
          "rc_expo": "0,0,0",
          "dterm_lowpass_hz_yaw": "90",
          "dterm_lowpass2_hz_pitch": "200",
          "P ratio": "64",
          "tpa_rate_d": "65",
          "gyro_hardware_lpf": "0",
          "airmode_activate_throttle": "32",
          "use_unsynced_pwm": "0",
          "tpa_rate_p": "75",
          "iterm_windup": "50",
          "spa_roll_d": "110",
          "gyro_notch_cutoff": "90,0",
        },
      ),
      stats: Stats(
        counts: FrameCounts(
          event: 1,
          main: 171694,
          slow: 11,
          gps: 0,
          gps_home: 0,
        ),
      ),
      capped: false,
      events: [
        End(
          disarm_reason: None,
        ),
      ],
      main: MainSnapshot(
        count: 171694,
        last_iteration: 692225,
        time: FieldSnapshot(
          min: 0,
          max: 106002824,
          unique: 171694,
          histogram: (10763, 10696, 10805, 10625, 10647, 10735, 10797, 10815, 10752, 10687, 10713, 10751, 10617, 10617, 10842, 10832),
        ),
        fields: Fields([
          FieldSnapshot(
            name: "axisP[0]",
            unit: Unitless,
            history: FieldSnapshot(
              min: -1581,
              max: 364,
              unique: 1038,
              histogram: (15763, 15551, 46687, 16520, 11019, 9666, 8729, 7563, 6591, 5869, 5326, 5040, 4607, 4407, 4235, 4121),
            ),
          ),
          FieldSnapshot(
            name: "axisP[1]",
            unit: Unitless,
            history: FieldSnapshot(
              min: -1308,
              max: 1969,
              unique: 1400,
              histogram: (66190, 31165, 20318, 13740, 9805, 6840, 5037, 3876, 2988, 2459, 1965, 1746, 1456, 1431, 1430, 1248),
            ),
          ),
          FieldSnapshot(
            name: "axisP[2]",
            unit: Unitless,
            history: FieldSnapshot(
              min: -513,
              max: 580,
              unique: 1092,
              histogram: (16392, 14854, 14456, 40676, 28865, 11341, 8669, 6968, 5892, 4991, 4099, 3687, 3349, 2814, 2539, 2102),
            ),
          ),
          FieldSnapshot(
            name: "axisI[0]",
            unit: Unitless,
            history: FieldSnapshot(
              min: -336,
              max: 347,
              unique: 684,
              histogram: (53067, 8362, 6681, 8154, 6385, 7647, 8043, 8550, 7739, 7949, 8007, 9767, 10035, 8435, 6942, 5931),
            ),
          ),
          FieldSnapshot(
            name: "axisI[1]",
            unit: Unitless,
            history: FieldSnapshot(
              min: -46,
              max: 263,
              unique: 310,
              histogram: (62554, 16454, 10490, 6597, 4656, 6064, 5103, 6217, 6007, 6191, 6414, 5031, 6331, 8856, 6425, 8304),
            ),
          ),
          FieldSnapshot(
            name: "axisI[2]",
            unit: Unitless,
            history: FieldSnapshot(
              min: -243,
              max: 174,
              unique: 418,
              histogram: (53835, 9985, 10587, 8771, 8526, 7545, 6811, 5301, 8225, 8037, 6574, 10268, 6143, 7419, 8040, 5627),
            ),
          ),
          FieldSnapshot(
            name: "axisD[0]",
            unit: Unitless,
            history: FieldSnapshot(
              min: -1808,
              max: 1924,
              unique: 711,
              histogram: (60640, 20516, 17815, 15537, 12452, 10101, 8092, 6387, 5060, 3793, 3068, 2336, 1913, 1614, 1304, 1066),
            ),
          ),
          FieldSnapshot(
            name: "axisD[1]",
            unit: Unitless,
            history: FieldSnapshot(
              min: -3205,
              max: 2667,
              unique: 916,
              histogram: (68135, 26832, 20602, 15544, 11154, 8066, 5537, 3937, 2937, 2181, 1767, 1345, 1175, 956, 830, 696),
            ),
          ),
          FieldSnapshot(
            name: "axisD[2]",
            unit: Unitless,
            history: FieldSnapshot(
              min: -98,
              max: 104,
              unique: 131,
              histogram: (149627, 15042, 3532, 1155, 727, 435, 372, 219, 96, 105, 106, 67, 65, 55, 52, 39),
            ),
          ),
          FieldSnapshot(
            name: "axisF[0]",
            unit: Unitless,
            history: FieldSnapshot(
              min: 0,
              max: 0,
              unique: 1,
            ),
          ),
          FieldSnapshot(
            name: "axisF[1]",
            unit: Unitless,
            history: FieldSnapshot(
              min: 0,
              max: 0,
              unique: 1,
            ),
          ),
          FieldSnapshot(
            name: "axisF[2]",
            unit: Unitless,
            history: FieldSnapshot(
              min: 0,
              max: 0,
              unique: 1,
            ),
          ),
          FieldSnapshot(
            name: "rcCommand[0]",
            unit: Unitless,
            history: FieldSnapshot(
              min: -176,
              max: 87,
              unique: 264,
              histogram: (8577, 4835, 6800, 23467, 71505, 9955, 8449, 7636, 3935, 5431, 3975, 3710, 4840, 3302, 2635, 2642),
            ),
          ),
          FieldSnapshot(
            name: "rcCommand[1]",
            unit: Unitless,
            history: FieldSnapshot(
              min: -137,
              max: 94,
              unique: 232,
              histogram: (42309, 56799, 15042, 7744, 4394, 4666, 4647, 4110, 2634, 4123, 3700, 2871, 6132, 5430, 3046, 4047),
            ),
          ),
          FieldSnapshot(
            name: "rcCommand[2]",
            unit: Unitless,
            history: FieldSnapshot(
              min: -122,
              max: 267,
              unique: 389,
              histogram: (11338, 7444, 5393, 8378, 64058, 19235, 7160, 6369, 5923, 4491, 4582, 5148, 6443, 5381, 5554, 4797),
            ),
          ),
          FieldSnapshot(
            name: "rcCommand[3]",
            unit: Unitless,
            history: FieldSnapshot(
              min: 0,
              max: 1522,
              unique: 508,
              histogram: (4605, 8631, 4890, 4557, 5136, 3318, 4654, 60042, 7286, 5508, 2885, 4644, 15282, 23080, 9243, 7933),
            ),
          ),
          FieldSnapshot(
            name: "setpoint[0]",
            unit: Unitless,
            history: FieldSnapshot(
              min: -94,
              max: 40,
              unique: 135,
              histogram: (10894, 19176, 87333, 6001, 3076, 4453, 5048, 3651, 4277, 6259, 6339, 3160, 4039, 3677, 2102, 2209),
            ),
          ),
          FieldSnapshot(
            name: "setpoint[1]",
            unit: Unitless,
            history: FieldSnapshot(
              min: -68,
              max: 44,
              unique: 113,
              histogram: (89435, 19146, 7268, 6160, 2902, 6616, 5272, 6319, 3321, 3295, 3337, 4060, 3926, 3969, 2647, 4021),
            ),
          ),
          FieldSnapshot(
            name: "setpoint[2]",
            unit: Unitless,
            history: FieldSnapshot(
              min: -59,
              max: 171,
              unique: 231,
              histogram: (8192, 6969, 81091, 10348, 5272, 6942, 8964, 9083, 5873, 5297, 4517, 3571, 3682, 5120, 4008, 2765),
            ),
          ),
          FieldSnapshot(
            name: "setpoint[3]",
            unit: Unitless,
            history: FieldSnapshot(
              min: 0,
              max: 523,
              unique: 505,
              histogram: (63241, 4078, 4478, 3038, 14067, 8048, 23561, 8975, 6048, 5996, 7987, 3487, 5118, 4364, 3669, 5539),
            ),
          ),
          FieldSnapshot(
            name: "gyroADC[0]",
            unit: Rotation,
            history: FieldSnapshot(
              min: -229,
              max: 974,
              unique: 728,
              histogram: (51902, 24393, 18498, 14378, 11382, 8737, 7236, 6148, 5284, 4806, 4117, 3587, 2975, 2786, 2756, 2709),
            ),
          ),
          FieldSnapshot(
            name: "gyroADC[1]",
            unit: Rotation,
            history: FieldSnapshot(
              min: -1060,
              max: 703,
              unique: 892,
              histogram: (57725, 24857, 13695, 9552, 7783, 7245, 6602, 5542, 5079, 4637, 4795, 5504, 5672, 5313, 4354, 3339),
            ),
          ),
          FieldSnapshot(
            name: "gyroADC[2]",
            unit: Rotation,
            history: FieldSnapshot(
              min: -303,
              max: 263,
              unique: 567,
              histogram: (55719, 15588, 12189, 10992, 10355, 9267, 7543, 6117, 6386, 6551, 6864, 6498, 5005, 4437, 4144, 4039),
            ),
          ),
          FieldSnapshot(
            name: "accSmooth[0]",
            unit: Acceleration,
            history: FieldSnapshot(
              min: -10898,
              max: 6599,
              unique: 2060,
              histogram: (10627, 9419, 9628, 9637, 10325, 10227, 9918, 9222, 9028, 8824, 9651, 13033, 13290, 13663, 13583, 11619),
            ),
          ),
          FieldSnapshot(
            name: "accSmooth[1]",
            unit: Acceleration,
            history: FieldSnapshot(
              min: -8977,
              max: 5637,
              unique: 1802,
              histogram: (8473, 9314, 10109, 10104, 10498, 9921, 10551, 12692, 14579, 19364, 12692, 10166, 8654, 8440, 8067, 8070),
            ),
          ),
          FieldSnapshot(
            name: "accSmooth[2]",
            unit: Acceleration,
            history: FieldSnapshot(
              min: -10158,
              max: 10461,
              unique: 3113,
              histogram: (8586, 8262, 8860, 11342, 15400, 18144, 15652, 11748, 9775, 9497, 9650, 9631, 9266, 8759, 8586, 8536),
            ),
          ),
          FieldSnapshot(
            name: "motor[0]",
            unit: Unitless,
            history: FieldSnapshot(
              min: 0,
              max: 1980,
              unique: 879,
              histogram: (8514, 8364, 8278, 8067, 8009, 8138, 7838, 7544, 7384, 7481, 7356, 7483, 7792, 11534, 48974, 8938),
            ),
          ),
          FieldSnapshot(
            name: "motor[1]",
            unit: Unitless,
            history: FieldSnapshot(
              min: 0,
              max: 1875,
              unique: 717,
              histogram: (5592, 5644, 6734, 14031, 19519, 11161, 6123, 5371, 5416, 5554, 5377, 5270, 5320, 18468, 46508, 5606),
            ),
          ),
          FieldSnapshot(
            name: "motor[2]",
            unit: Unitless,
            history: FieldSnapshot(
              min: 0,
              max: 1950,
              unique: 774,
              histogram: (7721, 7588, 7765, 7927, 8167, 7790, 7856, 7709, 7630, 7747, 8824, 7419, 7097, 18789, 38439, 13226),
            ),
          ),
          FieldSnapshot(
            name: "motor[3]",
            unit: Unitless,
            history: FieldSnapshot(
              min: 0,
              max: 1781,
              unique: 579,
              histogram: (5557, 5408, 5347, 5345, 5438, 5241, 5386, 5221, 5229, 4996, 5105, 5332, 5559, 56175, 40208, 6147),
            ),
          ),
        ]),
      ),
      slow: SlowSnapshot(
        count: 11,
        fields: Fields([
          FieldSnapshot(
            name: "flightModeFlags",
            unit: FlightMode,
            history: {
              "3D": 11,
              "ARM": 11,
            },
          ),
          FieldSnapshot(
            name: "stateFlags",
            unit: State,
            history: {},
          ),
          FieldSnapshot(
            name: "failsafePhase",
            unit: FailsafePhase,
            history: {
              "IDLE": 11,
            },
          ),
          FieldSnapshot(
            name: "rxSignalReceived",
            unit: Boolean,
            history: History(
              yes: 11,
              no: 0,
            ),
          ),
          FieldSnapshot(
            name: "rxFlightChannelsValid",
            unit: Boolean,
            history: History(
              yes: 11,
              no: 0,
            ),
          ),
        ]),
      ),
    )),
  ],
)
